syftlib/src/main/java/org/openmined/syft/networking/clients/SyftWebSocket.kt

```diff
@@ -109,31 +115,31 @@ internal class SyftWebSocket(
      * Create new web socket connection
      * */
     private fun connect() {
+        if (socketConnected.get()) {
+            return
+        }
+
         var retryDelay = 1000L
 
         compositeDisposable.add(
-            Completable.create {
-
-                if (socketConnected.get())
-                    return@create
-
+            socketStatusProcessor.retryWhen { errors ->
+                errors.zipWith(
+                    Flowable.range(1, MAX_RETRY_CONNECTS + 1),
+                    BiFunction<Throwable, Int, Int> { error: Throwable, retryCount: Int ->
+                        if (retryCount > MAX_RETRY_CONNECTS) throw error
+                        statusPublishProcessor.offer(NetworkMessage.SocketError(error))
+                        retryCount
+                    }
+                ).flatMap {
+                    retryDelay = min(retryDelay * 2, MAX_RETRY_TIMEOUT)
+                    Flowable.timer(retryDelay, TimeUnit.MILLISECONDS)
+                }
+            }.subscribe {
+                webSocket = it!!
                 socketConnected.set(true)
-                webSocket = client.newWebSocket(request, syftSocketListener)
```

Comment:
I think it should have been here only

Reply:
How?

Reply:
So why did you need to extract out new websocket? 
The flow should be like this, statusProcessor on success sets the `connected=True`  rather than it being set inside the `connect()``. Now in retry when after the timer for delay completes, it resubscribes so you should call the new websocket inside the `subscribe` block. Now whenever the socket times out onFailure will call `statusProcessor.onError()` making the socketProcessor to call retry when again.